[5.6] [TSVB] Fix memory leak

timroes opened the pull request:
## Summary

Summarize your PR. If it involves visual changes include a screenshot or gif.

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server)
- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request-legacy/63/)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request-legacy/66/)

timroes commented:
Diff to `5.6.16` release:

```diff
diff -urN src/core_plugins/metrics/server/lib/vis_data/helpers/index.js src/core_plugins/metrics/server/lib/vis_data/helpers/index.js
--- src/core_plugins/metrics/server/lib/vis_data/helpers/index.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/helpers/index.js	2020-05-11 16:59:03.000000000 +0200
@@ -48,6 +48,8 @@
 
 var _unit_to_seconds2 = _interopRequireDefault(_unit_to_seconds);
 
+var _overwrite = require('./overwrite');
+
 function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
 
 module.exports = {
@@ -62,5 +64,6 @@
   getTimerange: _get_timerange2.default,
   mapBucket: _map_bucket2.default,
   parseSettings: _parse_settings2.default,
-  unitToSeconds: _unit_to_seconds2.default
+  unitToSeconds: _unit_to_seconds2.default,
+  overwrite: _overwrite.overwrite
 };
diff -urN src/core_plugins/metrics/server/lib/vis_data/helpers/overwrite.js src/core_plugins/metrics/server/lib/vis_data/helpers/overwrite.js
--- src/core_plugins/metrics/server/lib/vis_data/helpers/overwrite.js	1970-01-01 01:00:00.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/helpers/overwrite.js	2020-05-11 16:59:03.000000000 +0200
@@ -0,0 +1,163 @@
+'use strict';
+
+Object.defineProperty(exports, "__esModule", {
+  value: true
+});
+exports.isPlainObject = isPlainObject;
+exports.set = set;
+exports.overwrite = overwrite;
+/*
+* Licensed to Elasticsearch B.V. under one or more contributor
+* license agreements. See the NOTICE file distributed with
+* this work for additional information regarding copyright
+* ownership. Elasticsearch B.V. licenses this file to you under
+* the Apache License, Version 2.0 (the "License"); you may
+* not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+function isObjectInner(val) {
+  return val != null && typeof val === 'object' && Array.isArray(val) === false;
+}
+
+function isObjectObject(o) {
+  return isObjectInner(o) === true && Object.prototype.toString.call(o) === '[object Object]';
+}
+
+function isPlainObject(o) {
+  if (isObjectObject(o) === false) return false;
+
+  // If has modified constructor
+  const ctor = o.constructor;
+  if (typeof ctor !== 'function') return false;
+
+  // If has modified prototype
+  const prot = ctor.prototype;
+  if (isObjectObject(prot) === false) return false;
+
+  // If constructor does not have an Object-specific method
+  if (prot.hasOwnProperty('isPrototypeOf') === false) {
+    return false;
+  }
+
+  // Most likely a plain Object
+  return true;
+}
+
+function isObject(val) {
+  return val !== null && (typeof val === 'object' || typeof val === 'function');
+}
+
+function createKey(pattern, options) {
+  let id = pattern;
+  if (typeof options === 'undefined') {
+    return id + '';
+  }
+  const keys = Object.keys(options);
+  for (let i = 0; i < keys.length; i++) {
+    const key = keys[i];
+    id += ';' + key + '=' + String(options[key]);
+  }
+  return id;
+}
+
+function isValidKey(key) {
+  return key !== '__proto__' && key !== 'constructor' && key !== 'prototype';
+}
+
+function set(target, path, value, options) {
+  if (!isObject(target)) {
+    return target;
+  }
+
+  const opts = options || {};
+  const isArray = Array.isArray(path);
+  if (!isArray && typeof path !== 'string') {
+    return target;
+  }
+
+  let merge = opts.merge;
+  if (merge && typeof merge !== 'function') {
+    merge = Object.assign;
+  }
+
+  const keys = (isArray ? path : split(path, opts)).filter(isValidKey);
+  const len = keys.length;
+  const orig = target;
+
+  if (!options && keys.length === 1) {
+    result(target, keys[0], value, merge);
+    return target;
+  }
+
+  for (let i = 0; i < len; i++) {
+    const prop = keys[i];
+
+    if (!isObject(target[prop])) {
+      target[prop] = {};
+    }
+
+    if (i === len - 1) {
+      result(target, prop, value, merge);
+      break;
+    }
+
+    target = target[prop];
+  }
+
+  return orig;
+}
+
+function result(target, path, value, merge) {
+  if (merge && isPlainObject(target[path]) && isPlainObject(value)) {
+    target[path] = merge({}, target[path], value);
+  } else {
+    target[path] = value;
+  }
+}
+
+function split(path, options) {
+  const id = createKey(path, options);
+  if (set.memo[id]) return set.memo[id];
+
+  const char = options && options.separator ? options.separator : '.';
+  let keys = [];
+  const res = [];
+
+  if (options && typeof options.split === 'function') {
+    keys = options.split(path);
+  } else {
+    keys = path.split(char);
+  }
+
+  for (let i = 0; i < keys.length; i++) {
+    let prop = keys[i];
+    while (prop && prop.slice(-1) === '\\' && keys[i + 1] != null) {
+      prop = prop.slice(0, -1) + char + keys[++i];
+    }
+    res.push(prop);
+  }
+  set.memo[id] = res;
+  return res;
+}
+
+set.memo = {};
+
+/**
+   * Set path in obj. Behaves like lodash `set`
+   * @param obj The object to mutate
+   * @param path The path of the sub-property to set
+   * @param val The value to set the sub-property to
+   */
+function overwrite(obj, path, val) {
+  set(obj, path, undefined);
+  set(obj, path, val);
+}
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/date_histogram.js src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/date_histogram.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/date_histogram.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/date_histogram.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,9 +5,7 @@
 });
 exports.default = dateHistogram;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
+var _helpers = require('../../helpers');
 
 var _get_bucket_size = require('../../helpers/get_bucket_size');
 
@@ -33,7 +31,7 @@
     const from = _getTimerange.from,
           to = _getTimerange.to;
 
-    _lodash2.default.set(doc, `aggs.${annotation.id}.date_histogram`, {
+    (0, _helpers.overwrite)(doc, `aggs.${annotation.id}.date_histogram`, {
       field: timeField,
       interval: intervalString,
       min_doc_count: 0,
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/top_hits.js src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/top_hits.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/top_hits.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/annotations/top_hits.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,17 +5,13 @@
 });
 exports.default = topHits;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
-
-function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
+var _helpers = require('../../helpers');
 
 function topHits(req, panel, annotation) {
   return next => doc => {
     const fields = annotation.fields && annotation.fields.split(/[,\s]+/) || [];
     const timeField = annotation.time_field;
-    _lodash2.default.set(doc, `aggs.${annotation.id}.aggs.hits.top_hits`, {
+    (0, _helpers.overwrite)(doc, `aggs.${annotation.id}.aggs.hits.top_hits`, {
       sort: [{
         [timeField]: { order: 'desc' }
       }],
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/date_histogram.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/date_histogram.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/date_histogram.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/date_histogram.js	2020-05-11 16:59:03.000000000 +0200
@@ -17,7 +17,7 @@
 
 var _get_interval_and_timefield2 = _interopRequireDefault(_get_interval_and_timefield);
 
-var _lodash = require('lodash');
+var _helpers = require('../../helpers');
 
 function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
 
@@ -38,7 +38,7 @@
           to = _offsetTime.to;
 
 
-    (0, _lodash.set)(doc, `aggs.${series.id}.aggs.timeseries.date_histogram`, {
+    (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.date_histogram`, {
       field: timeField,
       interval: intervalString,
       min_doc_count: 0,
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/filter_ratios.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/filter_ratios.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/filter_ratios.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/filter_ratios.js	2020-05-11 16:59:03.000000000 +0200
@@ -9,9 +9,7 @@
 
 var _bucket_transform2 = _interopRequireDefault(_bucket_transform);
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
+var _helpers = require('../../helpers');
 
 function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
 
@@ -21,10 +19,10 @@
   return () => doc => {
     if (series.metrics.some(filter)) {
       series.metrics.filter(filter).forEach(metric => {
-        _lodash2.default.set(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-numerator.filter`, {
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-numerator.filter`, {
           query_string: { query: metric.numerator || '*', analyze_wildcard: true }
         });
-        _lodash2.default.set(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-denominator.filter`, {
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-denominator.filter`, {
           query_string: { query: metric.denominator || '*', analyze_wildcard: true }
         });
 
@@ -38,13 +36,13 @@
               field: metric.field
             })
           };
-          _lodash2.default.set(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-numerator.aggs`, aggBody);
-          _lodash2.default.set(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-denominator.aggs`, aggBody);
+          (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-numerator.aggs`, aggBody);
+          (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}-denominator.aggs`, aggBody);
           numeratorPath = `${metric.id}-numerator>metric`;
           denominatorPath = `${metric.id}-denominator>metric`;
         }
 
-        _lodash2.default.set(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}`, {
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}`, {
           bucket_script: {
             buckets_path: {
               numerator: numeratorPath,
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/metric_buckets.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/metric_buckets.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/metric_buckets.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/metric_buckets.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,9 +5,7 @@
 });
 exports.default = metricBuckets;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
+var _helpers = require('../../helpers');
 
 var _get_bucket_size = require('../../helpers/get_bucket_size');
 
@@ -38,7 +36,7 @@
       if (fn) {
         try {
           const bucket = fn(metric, series.metrics, intervalString);
-          _lodash2.default.set(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}`, bucket);
+          (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.timeseries.aggs.${metric.id}`, bucket);
         } catch (e) {
           // meh
         }
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/sibling_buckets.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/sibling_buckets.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/sibling_buckets.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/sibling_buckets.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,9 +5,7 @@
 });
 exports.default = siblingBuckets;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
+var _helpers = require('../../helpers');
 
 var _get_bucket_size = require('../../helpers/get_bucket_size');
 
@@ -38,7 +36,7 @@
       if (fn) {
         try {
           const bucket = fn(metric, series.metrics, bucketSize);
-          _lodash2.default.set(doc, `aggs.${series.id}.aggs.${metric.id}`, bucket);
+          (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs.${metric.id}`, bucket);
         } catch (e) {
           // meh
         }
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_everything.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_everything.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_everything.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_everything.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,16 +5,12 @@
 });
 exports.default = splitByEverything;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
-
-function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
+var _helpers = require('../../helpers');
 
 function splitByEverything(req, panel, series) {
   return next => doc => {
     if (series.split_mode === 'everything' || series.split_mode === 'terms' && !series.terms_field) {
-      _lodash2.default.set(doc, `aggs.${series.id}.filter.match_all`, {});
+      (0, _helpers.overwrite)(doc, `aggs.${series.id}.filter.match_all`, {});
     }
     return next(doc);
   };
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filter.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filter.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filter.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filter.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,17 +5,13 @@
 });
 exports.default = splitByFilter;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
-
-function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
+var _helpers = require('../../helpers');
 
 function splitByFilter(req, panel, series) {
   return next => doc => {
     if (series.split_mode !== 'filter') return next(doc);
-    _lodash2.default.set(doc, `aggs.${series.id}.filter.query_string.query`, series.filter || '*');
-    _lodash2.default.set(doc, `aggs.${series.id}.filter.query_string.analyze_wildcard`, true);
+    (0, _helpers.overwrite)(doc, `aggs.${series.id}.filter.query_string.query`, series.filter || '*');
+    (0, _helpers.overwrite)(doc, `aggs.${series.id}.filter.query_string.analyze_wildcard`, true);
     return next(doc);
   };
 }
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filters.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filters.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filters.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_filters.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,18 +5,14 @@
 });
 exports.default = splitByFilter;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
-
-function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
+var _helpers = require('../../helpers');
 
 function splitByFilter(req, panel, series) {
   return next => doc => {
     if (series.split_mode === 'filters' && series.split_filters) {
       series.split_filters.forEach(filter => {
-        _lodash2.default.set(doc, `aggs.${series.id}.filters.filters.${filter.id}.query_string.query`, filter.filter || '*');
-        _lodash2.default.set(doc, `aggs.${series.id}.filters.filters.${filter.id}.query_string.analyze_wildcard`, true);
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.filters.filters.${filter.id}.query_string.query`, filter.filter || '*');
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.filters.filters.${filter.id}.query_string.analyze_wildcard`, true);
       });
     }
     return next(doc);
diff -urN src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_terms.js src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_terms.js
--- src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_terms.js	2019-03-13 16:45:11.000000000 +0100
+++ src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_terms.js	2020-05-11 16:59:03.000000000 +0200
@@ -5,9 +5,7 @@
 });
 exports.default = splitByTerm;
 
-var _lodash = require('lodash');
-
-var _lodash2 = _interopRequireDefault(_lodash);
+var _helpers = require('../../helpers');
 
 var _basic_aggs = require('../../../../../common/basic_aggs');
 
@@ -52,15 +50,15 @@
       const to = _getTimerange.to;
 
 
-      _lodash2.default.set(doc, `aggs.${series.id}.terms.field`, series.terms_field);
-      _lodash2.default.set(doc, `aggs.${series.id}.terms.size`, series.terms_size);
+      (0, _helpers.overwrite)(doc, `aggs.${series.id}.terms.field`, series.terms_field);
+      (0, _helpers.overwrite)(doc, `aggs.${series.id}.terms.size`, series.terms_size);
       const metric = series.metrics.find(item => item.id === series.terms_order_by);
       if (metric && metric.type !== 'count' && ~_basic_aggs2.default.indexOf(metric.type)) {
         const sortAggKey = `${series.terms_order_by}-SORT`;
         const fn = _bucket_transform2.default[metric.type];
         const bucketPath = (0, _get_buckets_path2.default)(series.terms_order_by, series.metrics).replace(series.terms_order_by, `${sortAggKey} > SORT`);
-        _lodash2.default.set(doc, `aggs.${series.id}.terms.order`, { [bucketPath]: 'desc' });
-        _lodash2.default.set(doc, `aggs.${series.id}.aggs`, {
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.terms.order`, { [bucketPath]: 'desc' });
+        (0, _helpers.overwrite)(doc, `aggs.${series.id}.aggs`, {
           [sortAggKey]: {
             filter: {
               range: {
```

timroes commented:
:heavy_check_mark: Done
